### PR TITLE
Fix mixins

### PR DESCRIPTION
**Pull request by krisselden**
WIP

### CODE REVIEW

**Comment by krisselden** [comment]
@stefanpenner I do want to prune toString, but it is orthogonal.

**Comment by krisselden** [comment]
unless that is copy toString is still a thing that is needed

**Comment by stefanpenner** [comment]
> unless that is copy toString is still a thing that is needed

i don't believe so

File: packages/ember-metal/lib/mixin.js [hunk 1]
```diff
@@ -221,12 +209,49 @@ function addNormalizedProperty(base, key, value, meta, descs, values, concats, m
...
 }
 
+function mixinProperties(mixinsMeta, mixin) {
+  var guid;
+
+  if (mixin instanceof Mixin) {
+    guid = guidFor(mixin);
+    if (mixinsMeta.peekMixins(guid)) {
+      // should be the same as undefined?
+      //return CONTINUE;
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: does this comment need to be addressed or is this for future travelers.

File: packages/ember-metal/lib/mixin.js [hunk 2]
```diff
@@ -221,12 +209,49 @@ function addNormalizedProperty(base, key, value, meta, descs, values, concats, m
   }
 }
 
+function mixinProperties(mixinsMeta, mixin) {
+  var guid;
+
+  if (mixin instanceof Mixin) {
+    guid = guidFor(mixin);
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: `let`

File: packages/ember-metal/lib/mixin.js [hunk 3]
```diff
@@ -55,20 +55,7 @@ function isMethod(obj) {
...
-  if (mixin instanceof Mixin) {
-    guid = guidFor(mixin);
-    if (mixinsMeta.peekMixins(guid)) { return CONTINUE; }
-    mixinsMeta.writeMixins(guid, mixin);
-    return mixin.properties;
-  } else {
-    return mixin; // apply anonymous mixin properties
-  }
-}
+var CONTINUE = {}; // remove
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: `const`

File: packages/ember-metal/lib/mixin.js [hunk 4]
```diff
@@ -362,6 +389,18 @@ function replaceObserversAndListeners(obj, key, observerOrListener) {
   }
 }
 
+// function elemEq(arr1, arr2) {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: ^ ?

File: packages/ember-metal/lib/mixin.js [hunk 5]
```diff
@@ -221,12 +209,49 @@ function addNormalizedProperty(base, key, value, meta, descs, values, concats, m
...
+  } else {
+    // TODO assert always is a Mixin
+    //console.assert(false, 'this was for createWithMixins');
+    return mixin; // apply anonymous mixin properties
+  }
+}
+
+function addNormalizedProperties(base, meta, props, allKeys, descs, values, concats, mergings) {
+  let keys = Object.keys(props);
+  for (let i = 0, l = keys.length; i < l; i++) {
```
Comment (stefanpenner) [inline packages/ember-metal/lib/mixin.js]: do we want to cache length?

**Comment by krisselden** [comment]
@stefanpenner I'm not sure about the keys.indexOf(key) !== -1 style unique-ing. I'm wondering about just doing Object.keys(values) again on the outside and just use a WITHOUT value to signal it should be skipped or do something to prevent without from inlining the delete or figure out something else with native array, maybe the mixin code be created more dynamically instead of using without after the fact

**Comment by rwjblue** [comment]
I'm somewhat confused where this is at.

@krisselden - Do you plan on pushing this forward (not rushing, just trying to see if we need to keep it open)?

**Comment by locks** [comment]
@krisselden is this still relevant?

**Comment by Serabe** [comment]
@krisselden should we keep this or just close it? Thank you!

**Comment by rwjblue** [comment]
I'll work on rebasing/updating this to get it current again.

**Comment by krisselden** [comment]
The main thing in here is making the keys array accurate so that we aren't using Object.keys() and `delete`

File: packages/ember-metal/lib/mixin.js (hunk 4 above)
Comment (krisselden) [inline packages/ember-metal/lib/mixin.js]: this was for the assert while I was developing to make sure the Object.keys(values) matched keys

**Comment by krisselden** [comment]
This would be a good PR to cleanup and merge, this is still a bad JIT issue

**Pull request closed without merging**